### DISCUSSION TITLE
Windows Forms Controls Inheritance!

### DISCUSSION THREAD

**Discussion by WaynePhillipsEA** [twinbasic/twinbasic#707] [Show and tell]
Ok so this is pretty cool... we can now do control _inheritance_ in twinBASIC, using the Implements-Via feature.

-  FIRST, make sure you've got the latest version of the compiler (v0.14.4 as of writing)

-  Start with the Sample 1a, add the `Windows Control DESIGNER Library` COM reference to your project.  (don't forget to save the Settings file and restart the compiler)

-  Add a new source file called `CustomCommandButton.twin`, and copy & paste this code into it:

```
[ WindowsControl ("/packages/WinNativeForms/miscellaneous/frmButton.png") ]
[ ClassId ("9EC2E963-75F8-4C0A-8E3D-BF8DEA49363B") ]
[ InterfaceId ("77F92188-BEB9-4E1A-A391-20FC9158BFA5") ]
[ COMCreatable (False) ]
Class CustomCommandButton
    
	Implements CommandButton Via BaseControl = New CommandButton()

	Public MyCustomProperty As String = "ChangeMe"
	
	Public Sub New(Serializer As WindowsControls.SerializationInfo)
		Serializer.Deserialize(Me)
	End Sub
	
	Private Sub OnInitialize() Handles BaseControl.Initialize
		Me.Caption = "Hello!"
	End Sub

	Private Sub OnClick() Handles BaseControl.Click
		MsgBox MyCustomProperty
	End Sub
	
End Class
```

That's it!  Open a form, and you should now find your CustomCommandButton available in the toolbox... and it has all the properties inherited from CommandButton, plus the `MyCustomProperty` you've added above.  When you click on the button at runtime, it should hit your OnClick handler.

You'll notice that the icon in the toolbox is the same as the normal CommandButton... but you can customize that by changing the path in the WindowsControl attribute... e.g. `[ WindowsControl("/miscellaneous/CustomCommandButton.png") ]` attribute in the above code, and dropping the custom icon into your projects miscellaneous folder.

Note that `Events` are NOT dealt with in the `Implements-Via` feature, so if you wanted to expose them, you'll have to expose them manually (by handling the BaseControl events and forwarding them to your own exposed Events).

You are of course not limited to adding just new properties... you can add new methods and events also.

**Comment by rboeckka** [comment]
Sorry, i need more information about: add this code: 
Do i need a new module, or in mainmodule  or in MyForm.twin?

**Reply by WaynePhillipsEA** [reply]
Thanks, I'll edit the op.   The code can be put in any twin file.   Best create a new one called CustomCommandButton.twin.

**Comment by WaynePhillipsEA** [comment]
Some of the attributes like `[Serialize]` and `[CustomDesigner]` are not correctly being picked up via Implements-Via here, so there's a couple of quirks with CustomCommandButton ... like the Caption property not appearing in the properties list, and the BackColor not having a color picker in the properties list.  These will be fixed soon.

**Reply by WaynePhillipsEA** [reply]
As of v0.14.6, this is now fixed, and so all inherited properties that come through the Implements-Via feature should appear correctly in the property sheet.

**Comment by WaynePhillipsEA** [comment]
Updated the OP to reflect changes in package v0.0.5, now available, that provides an `Initialize` event on all controls, giving you the opportunity to override properties and/or call base methods AFTER the control window handle is created and ready for use.
